srv/modules/runners/rebuild.py

```diff
@@ -0,0 +1,257 @@
+        log.info("Used: {} KB  Available: {} KB".format(used, available))
+        if used > available:
+            log.critical("OSDs exceed available free space")
+            return False
+        return True
+
+    def _busy_wait(self):
+        """ Wait until OSDs are safe to stop """
+        delay = 3
+        while True:
+            quiescent = self.local.cmd(self.master_minion,
+                                       'osd.ceph_quiescent', [], tgt_type="compound")
+            log.debug("quiescent: {}".format(quiescent))
+            if self.master_minion in quiescent and quiescent[self.master_minion] is True:
+                break
+            print("Waiting for PGs to recover...")
+            time.sleep(delay)
+            if delay < 60:
+                delay += 3
+
+    def _disengaged(self):
+        """ Return safety setting"""
+        return self.qrunner.cmd('disengage.check')
+
+    def _check_return(self, ret, minion):
+        """ Check for failures """
+        if isinstance(ret, str) or not all(ret.values()):
+            log.error("Failed to remove OSD(s)... skipping {}".format(minion))
+            self.skipped.append(minion)
+            return True
+        return False
+
+    def _check_deploy(self, deploy_ret, minion):
+        """ Check for failures """
+        log.debug(deploy_ret)
+        for ret in deploy_ret:
+            if minion in ret:
+                if ret[minion][0] != 0:
+                    pprint.pprint(ret[minion])
+                    self.skipped.append(minion)
+
+    def _skipped_summary(self):
+        """ Print summary of skipped minions """
+        if self.skipped:
+            print("The following minions were skipped:\n{}".format("\n".join(self.skipped)),
+                  "\n\nResolve any issues and run\n",
+                  "salt-run rebuild.nodes {}".format(" ".join(self.skipped)))
+
+    def _aggressive(self, osds, minion):
+        """ Delegate list to a single call """
+        osd_ret = self.runner.cmd(self.operation, osds)
+        log.debug("osd_ret: {}".format(osd_ret))
+
+        return self._check_return(osd_ret, minion)
+
+    def _check_each(self, osds, minion):
+        """ Wait between removal of each OSD """
+        for osd in osds:
+            osd_ret = self.runner.cmd(self.operation, [osd])
+            if self._check_return(osd_ret, minion):
+                return True
+            self._busy_wait()
+        return False
+
+    def run(self, checkonly=False):
+        """
+        For each minion
+           Retrieve the list of osds
+           Verify that the storage minion is safe to rebuild
+           Remove the OSDs
+           Deploy the minion
+        """
+        if not self._disengaged() and not checkonly:
+            log.error('Safety engaged...run "salt-run disengage.safety"')
+            return ""
+
+        for minion in self.minions:
+            log.info("minion: {}".format(minion))
```

Comment:
I think we should improve on the log.info output.

Reply:
The info is simply a bit more verbose of which minion is processed.

Reply:
my comment was addressing more the _format_ of the log output. reading just one word in the logs is not very informative imho.

Something like

> Processing minion: {} for foo/bar

is more pleasing to the eyes :>